src/navigation/partial_gate.py

```diff
@@ -0,0 +1,95 @@
+from dataclasses import dataclass
+from turtle import done
+import numpy as np
+
+from typing import Optional
+from state import BaseState
+from trajectory import Trajectory
+from drive import get_drive_command
+from aenum import Enum, NoAlias
+from context import Context
+
+STOP_THRESH = 0.2
+DRIVE_FWD_THRESH = 0.95
+POST_SEPARATION = 2  # meters
+
+
+@dataclass
+class PartialGateTrajectory(Trajectory):
+    def partial_gate_traj(post_pos, rover_pos):
+        """
+        Generates a trajectory to find second of two posts
+        :param post_pos:    position of the post (np.ndarray)
+        :param rover_pos:   position of the rover (np.ndarray). Assumes that the rover is facing the post
+        :return:            list of positions for the rover to traverse List(np.ndarray)
+        """
+        # The Path will be (assuming the vector <x, y> from rover to post
+        # Point one: left perpendicular, <-y,x>/mag * Post_separation
+        # Point two: opposite end, <x,y>/mag * Post_separation
+        # Point three: right perpendicular, <y,-x>/mag * Post_separation
+        # Essentially traveling around the circle made by the first post
+
+        rover_to_post = post_pos - rover_pos
+        rover_to_post *= POST_SEPARATION / np.linalg.norm(rover_to_post)
+        # scale vector to have magnitude == POST_SEPARATION
+
+        left_perp = np.array([-rover_to_post[1], rover_to_post[0], 0])  # (-y,x)\
```

Comment:
can you use the perpendicular2d function in np_utils.py?

Reply:
I can't really figure out how to correctly use this function - it looks like it takes a vector and is returning (-y, -x), but I might be parsing the function wrong

Reply:
whereas I want (y, -x) and (x, -y)

Reply:
hm is there a reason you actually want those values? At least by the variable name and what I can garner from your path, it seems like you want the vectors that are perpendicular to the vector towards the post. But what you're calculating is not perpendicular, you're calculating the vector mirrored over the x axis and the vector mirrored over the y axis which are only perpendicular if the angle of the original vector is 45 degrees. I think you actually want to do the perpendicular vectors for your path to be a square because right now I think it will probably be close enough to correct to work but I think you're actually getting like some other quadrilateral most of the time

Reply:
and if I misread the code or did the math long lmk I could've just totally misunderstood

Reply:
If the post is at (a, b) and rover is at (a + x, b + y), we'd calculate rover-to-post as c * (-x, -y). 

Then, 
- Point 1 (Left Perp): c * (a + y, b - x) 
- Point 2 (Diametrically Opposite): c * (a - x, b - y)
- Point 3 (Right Perp): c * (a - y, b + x)
- Point 4 (Originalish pos): c * (a + x, b + y)

1-3 Vector: c * (2y, -2x)
2-4 Vector: c * (-2x, -2y)

Dot product is 0 so they should be necessarily perpendicular — we also put it in a diagram here: https://github.com/umrover/mrover-ros/wiki/Navigation#GateTrajectory 

I think this will have to be a square?

We'll update the comments tho to be more detailed — we also added 4th point later and didn't update comments.

Reply:
Ah ok I see how this works. I think you guys could make this code a bit cleaner by doing everything relative to the post. Then you can use the perpendicular function and I think overall it will be more understandable. It would look something like this:

```
post_to_rover_normalized = normalized(rover - post)
 perp1 = perpendicular2d(post_to_rover_normalized) * c
 opposite = -1 * post_to_rover_normalized
 perp2 = -1 * perpendicular2d(post_to_rover_normalized)  * c
 finish = post_to_rover_normalized * c
 
 path = [perp1, opposite, perp2, finish]
 path_map_frame = [point + post_position for point in path_post_frame]
 ```

Reply:
wait actually sorry hold on

Reply:
I was tripping, I think what you're doing is fine

Reply:
looks like the perpendicular2d function is nputils is wrong tho which is actually pretty concerning lol

Reply:
should we fix that one too?

Reply:
hm hold off on that for now, I'm trying to figure out how the gate stuff was ever working with it wrong. But I'll get back to you asap

Reply:
but yes, eventually we should fix the function and just use the function here to make things a bit clearer

Reply:
alright updated the comments should be good now

Reply:
ok, so I went through and figured out what was going on with the gate and perpendicular 2d. I've updated perpendicular_2d to be correct so if you don't mind merging master into this pr and then using the perp2d function that'd be good

Reply:
sure no problem

Reply:
done — can't test on my machine at all, so would probably be good to run it once more in sim to make sure everything is still good.

Reply:
Just tested and fixed a little dimension thing - works the same as it did before (albeit a lil weird in sim bc of the post collision boxes)

Reply:
side note: apparently append() in numpy literally creates a new copy of the array which is absolutely disgusting. If there is a diff way to extend the dimension of a 2d np.array to a 3d np.array with a trailing 0 that is much preferred lol